node-js-basics-101/node_modules/socket.io-adapter/index.js

```diff
@@ -0,0 +1,263 @@
+    if (this.rooms[room].length === 0) delete this.rooms[room];
+  }
+
+  if (fn) process.nextTick(fn.bind(null, null));
+};
+
+/**
+ * Removes a socket from all rooms it's joined.
+ *
+ * @param {String} socket id
+ * @param {Function} callback
+ * @api public
+ */
+
+Adapter.prototype.delAll = function(id, fn){
+  var rooms = this.sids[id];
+  if (rooms) {
+    for (var room in rooms) {
+      if (this.rooms.hasOwnProperty(room)) {
+        this.rooms[room].del(id);
+        if (this.rooms[room].length === 0) delete this.rooms[room];
+      }
+    }
+  }
+  delete this.sids[id];
+
+  if (fn) process.nextTick(fn.bind(null, null));
+};
+
+/**
+ * Broadcasts a packet.
+ *
+ * Options:
+ *  - `flags` {Object} flags for this packet
+ *  - `except` {Array} sids that should be excluded
+ *  - `rooms` {Array} list of rooms to broadcast to
+ *
+ * @param {Object} packet object
+ * @api public
+ */
+
+Adapter.prototype.broadcast = function(packet, opts){
+  var rooms = opts.rooms || [];
+  var except = opts.except || [];
+  var flags = opts.flags || {};
+  var packetOpts = {
+    preEncoded: true,
+    volatile: flags.volatile,
+    compress: flags.compress
+  };
+  var ids = {};
+  var self = this;
+  var socket;
+
+  packet.nsp = this.nsp.name;
+  this.encoder.encode(packet, function(encodedPackets) {
+    if (rooms.length) {
+      for (var i = 0; i < rooms.length; i++) {
+        var room = self.rooms[rooms[i]];
+        if (!room) continue;
+        var sockets = room.sockets;
+        for (var id in sockets) {
+          if (sockets.hasOwnProperty(id)) {
+            if (ids[id] || ~except.indexOf(id)) continue;
+            socket = self.nsp.connected[id];
+            if (socket) {
+              socket.packet(encodedPackets, packetOpts);
+              ids[id] = true;
+            }
+          }
+        }
+      }
+    } else {
+      for (var id in self.sids) {
+        if (self.sids.hasOwnProperty(id)) {
+          if (~except.indexOf(id)) continue;
+          socket = self.nsp.connected[id];
+          if (socket) socket.packet(encodedPackets, packetOpts);
+        }
+      }
+    }
+  });
+};
+
+/**
+ * Gets a list of clients by sid.
+ *
+ * @param {Array} explicit set of rooms to check.
+ * @param {Function} callback
+ * @api public
+ */
+
+Adapter.prototype.clients = function(rooms, fn){
+  if ('function' == typeof rooms){
+    fn = rooms;
+    rooms = null;
+  }
+
+  rooms = rooms || [];
+
+  var ids = {};
+  var sids = [];
+  var socket;
+
+  if (rooms.length) {
+    for (var i = 0; i < rooms.length; i++) {
+      var room = this.rooms[rooms[i]];
+      if (!room) continue;
+      var sockets = room.sockets;
+      for (var id in sockets) {
+        if (sockets.hasOwnProperty(id)) {
+          if (ids[id]) continue;
+          socket = this.nsp.connected[id];
+          if (socket) {
+            sids.push(id);
+            ids[id] = true;
+          }
+        }
+      }
+    }
+  } else {
+    for (var id in this.sids) {
```

Comment:
__[jshint]__
* 'id' is already defined.

<!-- Comment by [SideCI](https://sideci.com) -->